VideoDecoder should not mandate codedWidth and codedHeight values

youennf opened the pull request:
#### d47d6dbc027edb2a805c6d4e584aa31df910ed01
<pre>
VideoDecoder should not mandate codedWidth and codedHeight values
<a href="https://bugs.webkit.org/show_bug.cgi?id=285743">https://bugs.webkit.org/show_bug.cgi?id=285743</a>
<a href="https://rdar.apple.com/142683375">rdar://142683375</a>

Reviewed by Jean-Yves Avenard.

WebCodecs config codedWidth and codedHeiht should only be used for selecting a decoder implementation, not for setting up the implementation itself.
Instead, the coded width and height should be retrieved from the stream itself or from the avcc/hvcc information.

We thus update H264, H265 implementation accordingly.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.js:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_av1-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_h264_annexb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_h264_avc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_h265_annexb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_h265_hevc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_vp8-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_vp9-expected.txt:
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
(-[RTCVideoDecoderH264 setAVCFormat:size:width:height:]):
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/RTCVideoDecoderH265.mm:
(spsAndVpsDataFromHvcc):
(ComputeH265ReorderSizeFromVPS):
(ComputeH265InfoFromHVCC):
(-[RTCVideoDecoderH265 setHVCCFormat:size:width:height:]):
(vpsDataFromHvcc): Deleted.
(ComputeH265ReorderSizeFromHVCC): Deleted.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/nalu_rewriter.cc:
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/nalu_rewriter.h:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
(WebKit::RemoteVideoDecoder::create):
(WebKit::RemoteVideoCodecFactory::createDecoder):
(WebKit::RemoteVideoDecoder::RemoteVideoDecoder):
(WebKit::RemoteVideoDecoder::decode):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::LibWebRTCCodecs::decodeFrame):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:

Canonical link: <a href="https://commits.webkit.org/288804@main">https://commits.webkit.org/288804@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08a3a088958e94aa1c4047f3339dc960056e94bf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84424 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4049 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38729 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89505 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35435 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4134 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12045 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65656 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23499 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87470 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3110 "Found 3 new test failures: fast/files/blob-stream-frame.html http/tests/websocket/tests/hybi/no-subprotocol.html http/wpt/cache-storage/quota-third-party.https.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76701 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45951 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3060 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30932 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34482 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73950 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31702 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90886 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11691 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8523 "Found 1 new test failure: http/tests/websocket/tests/hybi/websocket-allowed-setting-cookie-as-third-party.https.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74108 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11918 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72525 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73308 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18147 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17643 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16087 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3103 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11643 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17119 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11492 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14968 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13265 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -221,6 +221,12 @@ - (NSInteger)decodeData:(const uint8_t *)data
 }
 
 - (NSInteger)setAVCFormat:(const uint8_t *)data size:(size_t)size width:(uint16_t)width height:(uint16_t)height {
+  if (auto avcInformation = webrtc::ComputeH264InfoFromAVC(data, size)) {
```
should we error if we can't retrieve the information from the AVC? would mean it's invalid.

youennf replied:
I'll add a test with a bad AVC and will add some logging here.
I think VTB will likely error when submitting the frame.

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/RTCVideoDecoderH265.mm:
```diff
@@ -115,37 +118,52 @@ static void overrideColorSpaceAttachments(CVImageBufferRef imageBuffer) {
+            vpsData = { hvccData + position + hevcNalHeaderSize, size - hevcNalHeaderSize };
+            if (spsData.size())
+                return { spsData, vpsData };
         }
```
```suggested
        if (nalUnitType == webrtc::H265::NaluType::kSps)
          spsData = { hvccData + position + hevcNalHeaderSize, size - hevcNalHeaderSize };
        else if (nalUnitType == webrtc::H265::NaluType::kVps)
          vpsData = { hvccData + position + hevcNalHeaderSize, size - hevcNalHeaderSize };
        if (spsData.size() && vpsData.size())
          return { spsData, vpsData };
```

Intending seems incorrect for libwebRTC code (4 spaces vs 2 spaces)

youennf replied:
Yeah, we really need to move this code to WebCore.
Given this parsing routine is 4 space indented, I keep it like this.

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/RTCVideoDecoderH265.mm:
```diff
@@ -115,37 +118,52 @@ static void overrideColorSpaceAttachments(CVImageBufferRef imageBuffer) {
+    auto parsedVps = webrtc::H265VpsParser::ParseVps(vpsData, vpsDataSize);
     if (!parsedVps)
-        return 0;
+        return { };
```
it's supposed to return a uint8_t ; code before seems clearer

youennf replied:
Oh right, this was not supposed to be in this patch, will revert

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/RTCVideoDecoderH265.mm:
```diff
@@ -329,6 +347,14 @@ - (NSInteger)decodeData:(const uint8_t *)data size:(size_t)size timeStamp:(int64
 
 - (NSInteger)setHVCCFormat:(const uint8_t *)data size:(size_t)size width:(uint16_t)width height:(uint16_t)height {
+  if (auto hvccInformation = ComputeH265InfoFromHVCC(data, size)) {
+    if (hvccInformation->width)
```
is this ever going to be non-zero ?

youennf replied:
This can happen if the HVCC does not have a SPS, which would be strange but possible.

jyavenard reviewed (approved):
r=me with comments

youennf committed:
VideoDecoder should not mandate codedWidth and codedHeight values
https://bugs.webkit.org/show_bug.cgi?id=285743
rdar://142683375

Reviewed by Jean-Yves Avenard.

WebCodecs config codedWidth and codedHeiht should only be used for selecting a decoder implementation, not for setting up the implementation itself.
Instead, the coded width and height should be retrieved from the stream itself or from the avcc/hvcc information.

We thus update H264, H265 implementation accordingly.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any.js:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_av1-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_h264_annexb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_h264_avc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_h265_annexb-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_h265_hevc-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_vp8-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoDecoder-codec-specific.https.any_vp9-expected.txt:
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
(-[RTCVideoDecoderH264 setAVCFormat:size:width:height:]):
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/RTCVideoDecoderH265.mm:
(spsAndVpsDataFromHvcc):
(ComputeH265ReorderSizeFromVPS):
(ComputeH265InfoFromHVCC):
(-[RTCVideoDecoderH265 setHVCCFormat:size:width:height:]):
(vpsDataFromHvcc): Deleted.
(ComputeH265ReorderSizeFromHVCC): Deleted.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/nalu_rewriter.cc:
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/objc/components/video_codec/nalu_rewriter.h:
* Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
(WebKit::RemoteVideoDecoder::create):
(WebKit::RemoteVideoCodecFactory::createDecoder):
(WebKit::RemoteVideoDecoder::RemoteVideoDecoder):
(WebKit::RemoteVideoDecoder::decode):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::LibWebRTCCodecs::decodeFrame):
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:

Canonical link: https://commits.webkit.org/288804@main
